### PR TITLE
🐛 Disable out-of-bounds glyph placement

### LINKED ISSUE #700: invalid page, invisible custos? - Neon allows user to input glyphs outside of facsimile, results in invalid folio
In an issue possibly related to #651 and #697, a page turned up as invalid after some very trivial button click (sadly, I don't remember what). "Undo" created an apparently identical but valid page (see below)
image
image
As it turns out,  the MEI files for these two states is not identical; the "invalid" version has a custos on G at the end of the tenth staff,  where the "valid" one does not:
image
image

The zone for this otherwise invisible custos is outside the bounds of the page: 
image

### PR DESCRIPTION
**Pull request by jinh0**
Fixes #700 , at least for single glyphs / staves.

Now a notification will show up if you have placed a glyph out of bounds.

## Demo
https://user-images.githubusercontent.com/40512164/177211574-3e2e46cf-d6f5-484f-8d13-aa01cc2a8b2e.mov

### CODE REVIEW

File: src/utils/SelectTools.ts [hunk 1]
```diff
@@ -252,14 +252,7 @@ export function getStaffBBox (staff: SVGGElement): StaffBBox {
...
 
-  return {
-    id: staff.id,
-    ulx: ulx,
-    uly: uly,
-    lrx: lrx,
-    lry: lry,
-    rotate: rotate,
-  };
+  return { id: staff.id, ulx, uly, lrx, lry, rotate, };
```
Comment (jinh0) [inline src/utils/SelectTools.ts:255]: Shorten return with destructuring

File: src/utils/DragHandler.ts [hunk 2]
```diff
@@ -66,77 +65,55 @@ class DragHandler {
...
+        this.neonView.updateForCurrentPage();
+        this.endOptionsSelection();
+        this.reset();
+        this.dragInit();
+      });
   }
 
   /** Set the d3 action to use for [[reset]]. */
-  resetTo (reset: (selection: d3.Selection<d3.BaseType, {}, HTMLElement, any>, args: any[]) => void): void {
+  resetTo (reset: (selection: d3.Selection<d3.BaseType, unknown, HTMLElement, unknown>, args: unknown[]) => void): void {
```
Comment (jinh0) [inline src/utils/DragHandler.ts:116]: I had 3 warnings from ESlint every time I opened this file, so I finally just changed it to 'unknown'

File: src/utils/Coordinates.ts [hunk 3]
```diff
@@ -53,4 +60,39 @@ export function getStaffIdByCoords (clientX: number, clientY: number): string {
 export function getStaffByCoords (clientX: number, clientY: number): SVGGElement | null {
   const staffId = getStaffIdByCoords(clientX, clientY);
   return document.querySelector(`#${staffId}`);
-}
\ No newline at end of file
+}
+
+export function isOutOfSVGBounds (x: number, y: number): boolean {
+  const bgimg = document.querySelector<SVGImageElement>('#bgimg');
```
Comment (jinh0) [inline src/utils/Coordinates.ts:66]: bgImg

File: src/utils/Coordinates.ts [hunk 4]
```diff
@@ -53,4 +60,39 @@ export function getStaffIdByCoords (clientX: number, clientY: number): string {
...
+    x <= 0 || x >= Number(bgimg.getAttribute('width')) ||
+    y <= 0 || y >= Number(bgimg.getAttribute('height'))
+  );
+}
+
+/**
+ * Check whether the bounding box of an element is within the rectangle defined
+ * by the upper left and lower right points `ul` and `lr`
+ */
+export function isInside (bbox: BBox, ul: Point, lr: Point): boolean {
```
Comment (jinh0) [inline src/utils/Coordinates.ts:78]: specify point for ul and lr

File: src/utils/DragHandler.ts [hunk 5]
```diff
@@ -154,6 +131,105 @@ class DragHandler {
...
+
+    return isOutOfSVGBounds(endX, endY);
+  }
+
+  /**
+   * Returns whether the selection array is within the bounds of the SVG
+   * @param {SVGGraphicsElement[]} selection
+   * @returns {boolean} Is selection out of bounds
+   */
+  isSelOutOfBounds (selection: SVGGraphicsElement[]): boolean {
```
Comment (jinh0) [inline src/utils/DragHandler.ts:205]: rename to isDragOutOfBounds

File: src/utils/Coordinates.ts (hunk 4 above)
Comment (jinh0) [reply src/utils/Coordinates.ts:78]: dragUl and dragLr, and call the function isInsideDrag

File: src/utils/Select.ts [hunk 6]
```diff
@@ -97,6 +97,51 @@ function isSelByBBox (): boolean {
...
+
+  return getStaffBBox(el);
+}
+
+
+/**
+ * Checks whether an element is within the drag selection rectangle.
+ * Returns true if the element is within the bounds of `ul` (upper left) and `lr` (lower right)
+ */
+function isWithinRect(d: SVGGraphicsElement, ul: Point, lr: Point): boolean {
```
Comment (jinh0) [inline src/utils/Select.ts:132]: isElementInRect

**Review by yinanazhou — CHANGES_REQUESTED** [review]
Weird behaviour needs to be solved before merging:
https://user-images.githubusercontent.com/74484576/178404594-b5b1a3bf-f679-4640-9325-c5c002e77a9b.mov

**Review by yinanazhou — CHANGES_REQUESTED** [review]
Noticed another issue about dragging: the bbox is dragged along with the syllable.
This only happens in this repo, should be fixed in this PR.

https://user-images.githubusercontent.com/74484576/178646715-a5918e05-e31f-4cb9-854d-fb705c8b44f0.mov

**Comment by yinanazhou** [comment]
> Weird behaviour needs to be solved before merging: https://user-images.githubusercontent.com/74484576/178404594-b5b1a3bf-f679-4640-9325-c5c002e77a9b.mov

After discussion, this is now in issue #919, will be fixed in a separate PR.

**Review by yinanazhou — APPROVED** [review]

**Pull request merged**